gcs_light_client.mediawiki

```diff
@@ -0,0 +1,992 @@
+the fundamental components our set encoding relies on.
+
+
+=== Run-Length Encoding ===
+
+Run-Length Encoding (or RLE) is typically used in the video/image processing
+space to losslessly compresses images, or video frames. RLE works by ''omitting''
+the encoding of ''repeated'' values in a data stream. This achieves lossless
+compression as repeated items simply aren't transmitted. Instead, a value which
+represents the ''number of times'' a value repeats is transmitted.
+
+Typically RLE takes the form of encoding repeated values in a ''binary'' stream.
+A simple RLE scheme works as follows:
+* Encode the run length (number of occurrences) of 0's using <code>k</code> bits.
+** <code>k</code> acts as fixed length encoding for the length of a run.
+** This value acts as the maximum encodable run-length.
+* Transmission of runs of 1's is omitted.
+* Two 1's in a row are denoted by a zero-length run of zero.
+
+As an example, consider the following sequence of bits: 
+<pre>
+{0}^14 1 {0}^9 11 {0}^20 1 {0}^30 11 {0}^11
+</pre>
+
+The RLE of the bit stream above would be:  
+
+<pre>
+1110 1001 0000 1111 0101 1111 1111 0000 0000 1011
+</pre>
+
+RLE allows one to efficiently encode a data stream in a lossless manner. Due
+the the encoding of runs, RLE works best when encoding a set with a high degree
+of redundancy. A careful reader will notice that by using a fix-length encoding
+for the size of runs, efficiency is lost. Therefore, rather than using a
+fix-length encoding for the size of a run, we can instead use a ''variable''
+length encoding for the size of a run. This allows us to compress runs of a
+large size. To do so, we'll now turn to Golomb-Rice Coding.
+
+=== Golomb-Rice Coding ===
+
+RLE works well when encoding a data stream that has a high degree of redundancy.
+However, in our case due to the hashing of items within the compact filter,
+we'll be dealing with items that are ''uniformly distributed''. We can use this
+fact to leverage a more efficient encoding scheme based on the distribution of
+the length of a run. The [https://en.wikipedia.org/wiki/Geometric_distribution
+Geometric Distribution] represents the probabilities of a number of failures
+before the first success in a series of Bernoulli trials (yes/no experiments).
+If our values are i.i.d (independent, identically distributed) distributed of
+the run-length <code>r</code> can be represented as [6]:
+<pre>
+P(r = n) = p^n * (1-p)
+</pre>
+Intuitively, this calculates the probability of N zeroes (a run) followed by a
+single 1 (end of a run). Golomb coding takes advantage of this relationship to
+efficiently encode integers using a two-tuple. Given a group size of <code>m</code> one
+can encode an integer as:
+<pre>
+n = (q*m) + r
+  where q is (n / m)
+   and  r is n % m
+</pre>
+
+[https://en.wikipedia.org/wiki/Golomb_coding Golomb Coding] encodes the two
+values (<code>q</code> and <code>m</code> for a given integer <code>n</code> as a two-tuple. The first value
```

Comment:
typo: (q and *r*) for a given integer n

Also missing closing parenthesis.

Reply:
Fixed.